src/v/security/acl.cc

```diff
@@ -188,16 +227,27 @@ std::vector<std::vector<acl_binding>> acl_store::remove_bindings(
         // remove matching entries and track the deleted binding along with the
         // index of the filter that matched the entry.
         it->second.erase_if(
-          [&filters, &resource, &deleted, dry_run](const acl_entry& entry) {
+          [&filters, &resource, &deleted, &maybe_roles, dry_run](
+            const acl_entry& entry) {
               for (const auto& filter : filters) {
                   if (filter.first.entry().matches(entry)) {
                       auto binding = acl_binding(resource, entry);
-                      deleted.emplace(binding, filter.second);
+                      auto [it, _] = deleted.emplace(binding, filter.second);
+                      if (const auto& p = it->first.entry().principal();
+                          !dry_run && p.type() == principal_type::role) {
+                          maybe_roles.emplace_back(p);
+                      }
                       return !dry_run;
                   }
               }
               return false;
           });
+        for (const auto& principal : maybe_roles) {
+            it->second.remove_if_role(principal);
+        }
+        // ensure that elements won't outlive the corresponding bindings
+        // in enclosing scope.
+        maybe_roles.erase_to_end(maybe_roles.begin());
```

Comment:
Ah, my intention was to avoid releasing the memory, but it does look like this decays to `fragmented_vector::clear`

Reply:
maybe a `std::vector` is actually a better choice here?

Reply:
I don't have a good mental model of the limits here, but it feels safer to alloc a bit more other than crash (if this could grow to an oversized alloc).

Reply:
Oh, good spot - I hadn't noticed that it took the path of clear.

Reply:
>don't have a good mental model of the limits here

Neither do I, but I'd expect the number of iterations to be quite small in the common case. So potential savings is marginal compared to risk of the unknown. Probably just leave it as is for now.